src/pq_flash_index.cpp

```diff
@@ -343,37 +346,34 @@ namespace diskann {
           unsigned *nbrs = node_nhood + 1;
           // explore next level
           for (_u64 j = 0; j < nnbrs && !finish_flag; j++) {
-            if (std::find(node_list.begin(), node_list.end(), nbrs[j]) ==
-                node_list.end()) {
+            if (node_set.find(nbrs[j]) == node_set.end()) {
               cur_level->insert(nbrs[j]);
             }
-            if (cur_level->size() + node_list.size() >= num_nodes_to_cache) {
+            if (cur_level->size() + node_set.size() >= num_nodes_to_cache) {
               finish_flag = true;
             }
           }
           aligned_free(nhood.second);
         }
       }
 
-      diskann::cout << ". #nodes: " << node_list.size() - prev_node_list_size
+      diskann::cout << ". #nodes: " << node_set.size() - prev_node_set_size
                     << ", #nodes thus far: " << node_list.size() << std::endl;
-      prev_node_list_size = node_list.size();
+      prev_node_set_size = node_set.size();
       lvl++;
     }
 
-    std::vector<uint32_t> cur_level_node_list;
-    for (const unsigned &p : *cur_level)
-      cur_level_node_list.push_back(p);
-
-    std::shuffle(cur_level_node_list.begin(), cur_level_node_list.end(), urng);
-    size_t residual = num_nodes_to_cache - node_list.size();
+    assert(node_set.size() + cur_level->size() == num_nodes_to_cache ||
+           cur_level->size() == 0);
 
-    for (size_t i = 0; i < (std::min)(residual, cur_level_node_list.size());
-         i++)
-      node_list.push_back(cur_level_node_list[i]);
+    node_list.clear();
```

Comment:
nit: since size is known, can we reserve the size on the `node_list `first?

Reply:
done